### PR TITLE
feat: Reference Consistency Mask for Qwen-Image-Edit

### PR DESCRIPTION
**Pull request by kohya-ss**
This pull request adds support for inpainting and Reference Consistency Mask (RCM) features to the Qwen-Image(-Edit) generation script, along with new command-line options and prompt-line overrides for these features. It also introduces a bell notification option, improves handling of edit modes, and makes several related usability improvements.

This also fixes the issue where 2509 could not be inferred in batch mode and the issue where the control image was resized even if `--resize_control_to_image_size` was not specified.

**New inpainting and RCM features:**

* Added support for inpainting via a mask image (`--mask_path`) and implemented logic to apply the mask during the denoising loop for Qwen-Image-Edit, including prompt-line override (`--m`). [[1]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494R88-R93) [[2]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494L687-R784) [[3]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494R829-R890) [[4]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494L196-R242)
* Added Reference Consistency Mask (RCM) feature, with new arguments for threshold, relative threshold, kernel size, dilation, and debug saving (`--rcm_threshold`, `--rcm_relative_threshold`, `--rcm_kernel_size`, `--rcm_dilate_size`, `--rcm_debug_save`), and corresponding prompt-line overrides. The RCM logic is applied during the denoising loop for Qwen-Image-Edit. [[1]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494R148-R174) [[2]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494L687-R784) [[3]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494R829-R890) [[4]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494L196-R242)

**Usability improvements:**

* Added `--bell` option to ring a bell sound when generation is complete (supports both batch and interactive modes). [[1]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494R148-R174) [[2]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494R1296-R1298) [[3]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494R1395-R1397) [[4]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494R1418-R1420)
* Improved detection and handling of edit modes (`edit` or `edit_plus`) throughout the script, ensuring correct logic for Qwen-Image-Edit features. [[1]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494L960-R1110) [[2]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494L978-R1128) [[3]](diffhunk://#diff-fdad98d44e0f66647811cf938dc316f3cbf2df9b23de79358e3eaa2cb1c0c494L1182-R1335)

**Code and documentation fixes:**

* Fixed documentation for `pack_latents` to accurately describe the latent shape transformation.
* Updated scheduler logging and comments to reflect the correct scheduler class name (`FlowMatchEulerDiscreteScheduler`).
* Added missing imports for image and tensor processing (`torch.nn.functional`, `torchvision.transforms.functional`, `PIL.Image`).

### CODE REVIEW

**Comment by kohya-ss** [comment]
A more technical explanation created by Gemini 2.5 Pro.

---

### Description

This pull request introduces a new feature called **Reference Consistency Mask (RCM)** to address a common issue in Qwen-Image-Edit where the generated image has a slight positional drift or misalignment compared to the control image. RCM significantly improves the structural stability and positional accuracy of the image editing process.

This feature is implemented based on the idea of dynamically creating a mask during the denoising loop to "anchor" the parts of the image that should remain consistent with the reference (control) image.

### How RCM Works

For each step in the denoising loop, RCM performs the following actions:
1.  It calculates a "noisy" version of the original control latent, corresponding to the current timestep `t`.
2.  It computes the difference between the current generation latent and the noisy control latent.
3.  Areas with a small difference are considered "consistent" and are masked to be preserved. The sensitivity is controlled by the `rcm_threshold`.
4.  This mask is then used to reset the consistent regions of the current latent back to the state of the noisy reference latent, just before the `scheduler.step` is called.

This self-correcting mechanism prevents the accumulation of positional errors throughout the denoising process, ensuring that unchanged elements like backgrounds or faces stay perfectly aligned.

### New Features and Options

This PR adds the following command-line arguments to control RCM:

*   `--rcm_threshold`: Enables RCM and sets the threshold for detecting changes. Lower values result in a larger "edited" region. (e.g., `0.05`)
*   `--rcm_relative_threshold`: If set, the threshold becomes relative to the maximum difference found in the current step, making it more adaptive to different images and timesteps.
*   `--rcm_kernel_size`: Applies a Gaussian blur to the latents before comparison to create a smoother, more coherent mask. Defaults to `3`.
*   `--rcm_dilate_size`: Dilates the "edited" region of the mask to prevent artifacts at the boundaries of the edited area. Defaults to `0` (disabled).
*   `--rcm_debug_save`: Saves the generated mask for each step as a PNG image, which is extremely useful for tuning the parameters.

Additionally, this PR integrates support for traditional, user-provided inpainting masks via the `--mask_path` argument.

### Important Usage Notes

-   **Compatibility:** Both RCM and the standard inpainting mask are only effective in **edit mode** (when a control image is provided).
-   **Requirement:** To use these features, the initial control image must have the **same dimensions** as the final output image. The script will show an error and disable RCM if the sizes do not match.
-   **Exclusivity:** RCM and `--mask_path` cannot be used at the same time.
-   **Debugging Tip:** When first using RCM, it is highly recommended to use the `--rcm_debug_save` flag. This will save the masks to the output directory, allowing you to visually inspect how the `threshold` and other parameters are affecting the mask generation.

This implementation should provide users with a powerful new tool for achieving high-fidelity, positionally-stable image edits with the Qwen-Image-Edit model.

<details>
<summary>日本語</summary>

### 概要

このPull Requestは、**Reference Consistency Mask (RCM)** と呼ばれる新機能を導入します。これは、Qwen-Image-Editにおいて、生成画像が制御画像と比較してわずかな位置ずれを起こすという一般的な問題を解決するためのものです。RCMは、編集プロセスにおける構造的な安定性と位置精度を大幅に向上させます。

この機能は、denoisingループ中に動的にマスクを生成し、参照元（制御画像）と一致すべき部分を「固定（アンカー）」するというアイデアに基づいています。

### RCMの仕組み

RCMは、denoisingループの各ステップで以下の処理を実行します。
1.  現在のタイムステップ`t`に対応する、元の制御画像の潜在変数にノイズを加えたバージョンを計算します。
2.  現在の生成中latentと、ノイズ付加済み制御latentとの差分を計算します。
3.  差分が小さい領域を「一致している」とみなし、その部分を保持するようにマスクします。この感度は`rcm_threshold`によって制御されます。
4.  そして、このマスクを使い、`scheduler.step`が呼び出される直前に、一致している領域をノイズ付加済み参照latentの状態にリセットします。

この自己修正的なメカニズムにより、denoisingプロセス全体を通して位置誤差が蓄積されるのを防ぎ、背景や顔のような変更しない要素が完全に位置ずれなく維持されることを保証します。

### 新機能とオプション

このPRでは、RCMを制御するために以下のコマンドライン引数を追加します。

*   `--rcm_threshold`: RCMを有効にし、変更を検出するための閾値を設定します。値が低いほど、「編集される」領域が広くなります。（例: `0.05`）
*   `--rcm_relative_threshold`: このオプションを指定すると、閾値が現在のステップで見つかった差分の最大値に対する相対値となり、様々な画像やタイムステップに対してより適応的に動作します。
*   `--rcm_kernel_size`: 比較前にlatentにガウシアンブラーを適用し、よりスムーズで一貫性のあるマスクを生成します。デフォルトは`3`です。
*   `--rcm_dilate_size`: マスクの「編集される」領域を膨張させ、編集領域の境界部分に発生するアーティファクトを防ぎます。デフォルトは`0`（無効）です。
*   `--rcm_debug_save`: 各ステップで生成されたマスクをPNG画像として保存します。これはパラメータを調整する際に非常に役立ちます。

さらに、このPRは`--mask_path`引数による、ユーザーが提供する従来のinpaintingマスクのサポートも統合しています。

### 重要な使用上の注意

-   **互換性:** RCMと標準のinpaintingマスクは、どちらも**Editモード**（制御画像が提供されている場合）でのみ有効です。
-   **要件:** これらの機能を使用するには、最初の制御画像が最終的な出力画像と**同じサイズ**である必要があります。サイズが一致しない場合、スクリプトはエラーを表示し、RCMを無効にします。
-   **排他性:** RCMと`--mask_path`は同時に使用できません。
-   **デバッグのヒント:** 初めてRCMを使用する際は、`--rcm_debug_save`フラグを使用することを強く推奨します。これによりマスクが出力ディレクトリに保存され、`threshold`などのパラメータがマスク生成にどのように影響しているかを視覚的に確認できます。

この実装は、Qwen-Image-Editモデルで忠実度が高く、位置的に安定した画像編集を実現するための、強力な新しいツールをユーザーに提供するはずです。

</details>

**Comment by kohya-ss** [comment]
Technical limitations: Large drifts can lead to inconsistent generated images at the mask boundaries.

The following prompt options have been added:

- `--m`: mask_path
- `--rcm_th`: rcm_threshold
- `--rcm_rel_th`: rcm_relative_threshold, `--rcm_rel_th 1` or `--rcm_rel_th True` enables this, `--rcm_rel_th 0` or `--rcm_rel_th False` disbles this.
- `--rcm_ks`: rcm_kernel_size
- `--rcm_ds`: rcm_dilate_size

Sample: Prompt "Dress the girl in a school uniform.", 20 steps, seed 1234.

**Control image:**
ori1_tachi2

**Without RCM:**


**With RCM:** rcm_threshold=0.2, rcm_relative_threshold enabled, rcm_kernel_size 3,  rcm_dilation_size 2

File: src/musubi_tuner/qwen_image_generate_image.py [hunk 1]
```diff
@@ -684,17 +721,67 @@ def generate(
...
     assert n == num_inference_steps, f"Expected steps={num_inference_steps}, got {n} from scheduler."
 
-    num_warmup_steps = 0  # because FlowMatchingDiscreteScheduler.order is 1, we don't need warmup steps
+    num_warmup_steps = 0  # because FlowMatchEulerDiscreteScheduler.order is 1, we don't need warmup steps
 
     # handle guidance
     guidance = None  # guidance_embeds is false for Qwen-Image
 
+    # inpainting mask and RCM
+    has_same_size_control = is_edit and len(img_shapes[0]) > 1 and img_shapes[0][1] == img_shapes[0][0]
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/qwen_image_generate_image.py:738]: `img_shapes[0][0]` is the shape of the generation image, and `img_shapes[0][1]` is the shape of the control image, so this is fine.

File: src/musubi_tuner/qwen_image_generate_image.py [hunk 2]
```diff
@@ -739,6 +826,68 @@ def generate(
                 noise_norm = torch.norm(comb_pred, dim=-1, keepdim=True)
                 noise_pred = comb_pred * (cond_norm / noise_norm)
 
+            # Reference Consistency Mask (RCM) or inpainting
+            if rcm_enabled and i > 0 or inpainting_mask is not None:
+                t_n = t / 1000
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/qwen_image_generate_image.py:835]: `1000` is already used other places, so this will be fine.

File: src/musubi_tuner/qwen_image_generate_image.py [hunk 3]
```diff
@@ -739,6 +826,68 @@ def generate(
...
+                    diff_per_channel = diff.mean(dim=1, keepdim=True)  # mean over channels, (B, 1, h, w)
+                    th = (
+                        args.rcm_threshold
+                        if not args.rcm_relative_threshold
+                        else args.rcm_threshold * diff_per_channel.max().item()
+                    )
+                    latent_mask = (diff_per_channel >= th).to(dtype=torch.bfloat16)  # 1.0 for inpainting region, 0.0 for keep
+
+                    # resize mask to match packed latent size (H//16, W//16)
+                    latent_mask = F.interpolate(latent_mask, size=(height // 16, width // 16), mode="bilinear", align_corners=False)
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/qwen_image_generate_image.py]: Fixed.

File: src/musubi_tuner/qwen_image_generate_image.py [hunk 4]
```diff
@@ -739,6 +826,68 @@ def generate(
...
+                    if args.rcm_dilate_size > 0:
+                        # Use max pooling for dilation, which expands the area while keeping the mask smooth.
+                        latent_mask = F.max_pool2d(
+                            latent_mask, kernel_size=args.rcm_dilate_size * 2 + 1, stride=1, padding=args.rcm_dilate_size
+                        )
+
+                    # visualize mask for debugging, H*W -> H, W
+                    if args.rcm_debug_save:
+                        mask_np = (latent_mask[0, :].reshape(int(height / 16), int(width / 16)).float().cpu().numpy() * 255).astype(
+                            np.uint8
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/qwen_image_generate_image.py]: Fixed.

**Comment by sdbds** [comment]
Given that we have already added the mask_path feature, should we also port over the alpha_mask feature?
